### ISSUE TITLE
mocha as canonical runner for ethereal unit tests

### ISSUE THREAD

**Issue by zomglings** [the-chaingang/ethereal#24]
# Proposal: Use `mocha` as the canonical runner for `ethereal` unit tests

## General considerations

As of the time of writing of this proposal, `ethereal` does not have unit tests on the `master` branch. However, we do have a proposal for the structure of the `ethereal` testing interface: https://github.com/the-chaingang/ethereal/issues/20

We also have a corresponding PR which implements that interface: https://github.com/the-chaingang/ethereal/pull/23

Although these represent specific proposals for the interface, there are a few things that we can assert about any possible testing interface: 

1. A single test or suite of tests will consist of bringing up an `ethereal` network, running test code on the network once the nodes are all up, and optionally tearing down the network when the tests have run.

1. `docker-compose` is the framework through which the `ethereal` network in question will be brought up. This is a decision that will not change for a long time.

1. Each (suite of) test(s) should be brought up in separate `docker-compose` processes. Doing otherwise would lead to serious performance issues as the number of tests increases (number of docker processes running geth nodes grows at least linearly in number of test suites).

1. This means that each test should live in its own directory, which will contain documentation about that specific test as well as the fixtures required to run the test.

1. `docker-compose` offers a tear down primitive in `docker-compose down -v`.

1. This means that, as much as possible, we should be doing all our work through `docker-compose`, which does not offer a REST API or client libraries in any language - we are going to have to manually run `docker-compose` processes through our test runner.

## What problem is this proposal trying to solve?

Given the discussion above, our automated testing framework has to coordinate the execution of the various `docker-compose` threads that different test suites will run in, and also organize the teardown step for each test suite (which will involve destruction of volumes and networks, for example).

There are many frameworks that provide this functionality. Which one should we choose in `ethereal`?

## Candidate solutions

For the language our test runner is built for, let us consider the languages that are currently popular in the `ethereum` ecosystem, as our testing framework will be used by `ethereal` users in their simulation use cases in addition to being used by us to test our code. These languages are:

1. Javascript in general (with [node.js](https://nodejs.org/en/) being the particular flavor of Javascript that would suit our purposes well here)

1. [go](https://golang.org/)

1. [Python](https://www.python.org/)

1. [C++](https://en.wikipedia.org/wiki/C%2B%2B)

Assuming that the developer familiarity statistics for blockchain developers will soon roughly resemble (if they do not already) those of the general population of developers, it is clear that we should be betting on node.js test frameworks for our test runner [Reference: [Stackoverflow 2017 survey](https://insights.stackoverflow.com/survey/2017#technology)].

Within the Javascript ecosystem, [mocha](https://mochajs.org/) is the clear winner. [Reference: [The State of Javascript 2017](https://stateofjs.com/2017/testing/results)]

The facts that `mocha` is unopinionated, that it doesn't get in your way as a developer, and that all three of the current active contributors to this project are familiar with it just add to its allure.

## Proposal

Let us use `mocha` to run our unit tests. Besides being a powerful and flexible testing framework, it is also most likely that an incoming contributor will be familiar with it over other testing frameworks.

**Comment by zomglings** [comment]
@olawiberg , @shad0wfax : This is ready to review, please take a look.

**Comment by olawiberg** [comment]
@nkashy1 I agree that focusing on implementing the tests using `mocha` is the best starting point as there is a big community already using Javascript tools, such as Truffle and testrpc, for Ethereum development. 

Let's go ahead and implement this as the next step for running tests.

**Comment by shad0wfax** [comment]
My main reservation is that this forces the commiters on the Ethereal repository to have node and mocha installed (and potentially some could have some reservations using them). 
However given how difficult it is to find a universal, low friction language to build this on, we might as well go with whatever is lightweight and simple. For that reason, I second using mocha.
